dataforest/utils/catalogue.py

```diff
@@ -0,0 +1,23 @@
+import logging
+
+import pandas as pd
+
+
+def run_id_from_multi_row(run_id_rows: pd.DataFrame) -> str:
+    """
+    Extract the `run_id` from a `run_catalogue` where multiple rows match the
+    given `RunSpec`. Note that this shouldn't happen, but in the case of a
+    double entry, or two different `run_id`s for the same `RunSpec`, this gets
+    the first.
+    Args:
+        run_id_rows: rows of run_catalogue matching a given `RunSpec`
+    """
+    unique = list(run_id_rows["run_id"].unique())
+    run_id = unique[0]
+    if len(unique) != 1:
```

Comment:
This only gets called if more than one row of the catalogue matches the spec used to query it. So when you load the catalogue, you have a dataframe with a string representation of the spec as the index, and a single column containing the run ids. And when you slice the index with the current spec, you should just get one row. But in the case you don't (e.g. duplicates caused by some bug -- more likely once we get S3 cloud sync set up), this function runs to just pick the first matching row. If that doesn't make sense, or isn't what you were getting at, Slack me, otherwise, feel free to resolve it.

Reply:
Oh I see now. So `run_id_rows` are already sliced based on `RunSpec`, and for the same `RunSpec`, we must have the same `RunId`. If there are duplicates, we'll just get the first instance. If there are different IDs for the same spec then smth is wrong with hashing?

Reply:
Yeah, that's right. If there are duplicates with the same hash, then one run just got catalogued twice, somehow. If there are duplicates with different hashes, it could mean, for example, that two people pulled a branch from S3 and ran the same thing locally, each of which generating a different hash, then both pushed back up. Ideally, the cloud sync would handle appropriately, but this is just in case.

Reply:
This example makes a lot of sense now, thanks! 😃